core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,328 @@
+    appendA.appendFile(FILE_A).apply();
+
+    table.updateProperties()
+        .set(TableProperties.COMMIT_NUM_RETRIES, "0")
+        .commit();
+
+    table.ops().failCommits(1);
+
+    AssertHelpers.assertThrows("Should reject commit",
+        CommitFailedException.class, "Injected failure",
+        () -> table.newFastAppend().appendFile(FILE_B).commit());
+
+    table.updateProperties()
+        .set(TableProperties.COMMIT_NUM_RETRIES, "5")
+        .commit();
+
+    appendA.commit();
+
+    Assert.assertEquals(1, table.currentSnapshot().sequenceNumber());
+
+    AppendFiles appendFiles = table.newFastAppend().appendFile(FILE_C);
+    appendFiles.apply();
+    table.newFastAppend().appendFile(FILE_D).commit();
+    appendFiles.commit();
+
+    validateDataFiles(files(FILE_A, FILE_D, FILE_C), seqs(1, 2, 3));
+  }
+
+  @Test
+  public void testRollBack() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        2, table.currentSnapshot().sequenceNumber());
+
+    table.manageSnapshots().rollbackTo(snapshotId).commit();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        1, table.currentSnapshot().sequenceNumber());
+
+    Assert.assertEquals("Table last sequence number should be 2",
+        2, table.operations().current().lastSequenceNumber());
+
+    table.newFastAppend().appendFile(FILE_C).commit();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        3, table.currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testSingleTransaction() {
+    Transaction txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_A).commit();
+    txn.commitTransaction();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        1, table.currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testConcurrentTransaction() {
+    Transaction txn1 = table.newTransaction();
+    Transaction txn2 = table.newTransaction();
+    Transaction txn3 = table.newTransaction();
+    Transaction txn4 = table.newTransaction();
+
+    txn1.newFastAppend().appendFile(FILE_A).commit();
+    txn3.newOverwrite().addFile(FILE_C).commit();
+    txn4.newDelete().deleteFile(FILE_A).commit();
+    txn2.newAppend().appendFile(FILE_B).commit();
+
+    txn1.commitTransaction();
+    txn2.commitTransaction();
+    txn3.commitTransaction();
+    txn4.commitTransaction();
+
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        4, table.currentSnapshot().sequenceNumber());
+    validateDataFiles(files(FILE_A, FILE_B, FILE_C), seqs(4, 2, 3));
+  }
+
+  @Test
+  public void testMultipleOperationsTransaction() {
+    Transaction txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_A).commit();
```

Comment:
The sequence number of FILE_A needs to be validated, along with the rest of the metadata for this commit. Transactions will still produce a snapshot for every operation, it will just swap the existing state for the final state.